ffi_internals/src/native_type_data.rs

```diff
@@ -105,6 +114,154 @@ impl From<(NativeType, WrappingType)> for NativeTypeData {
             option: wrapping_type == WrappingType::Option
                 || wrapping_type == WrappingType::OptionVec,
             vec: wrapping_type == WrappingType::Vec || wrapping_type == WrappingType::OptionVec,
+            result: false,
+        }
+    }
+}
+
+/// Describes the initial state when parsing a `syn::Type`, where we have not yet determined
+/// whether the underlying type is wrapped in an `Option`, `Vec`, or `Result`.
+///
+/// This is basically an intermediary type to make it easier to get to `NativeTypeData`. Usage
+/// should look something like this:
+/// ```
+/// use quote::format_ident;
+/// use ffi_internals::native_type_data::{UnparsedNativeTypeData, NativeTypeData, NativeType};
+///
+/// let ty: syn::Type = syn::parse_str("Result<Foo>").unwrap();
+/// let initial = UnparsedNativeTypeData::initial(ty);
+/// let native_type_data = NativeTypeData::from(initial);
+/// assert_eq!(native_type_data.native_type, NativeType::Boxed(format_ident!("Foo")));
+/// assert_eq!(native_type_data.result, true);
+/// assert_eq!(native_type_data.option, false);
+/// assert_eq!(native_type_data.vec, false);
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct UnparsedNativeTypeData {
+    /// The type being parsed.
+    pub ty: Type,
+    /// Whether `ty` was discovered inside of an `Option`.
+    pub option: bool,
+    /// Whether `ty` was discovered inside of a `Vec`, `Array`, or slice.
+    pub vec: bool,
+    /// Whether `ty` was discovered in the `Success` variant of a `Result`.
+    pub result: bool,
+}
+
+impl UnparsedNativeTypeData {
+    /// The initial state for `UnparsedNativeTypeData`, where the `option`, `vec` and `result`
+    /// fields are all set to false.
+    ///
+    pub fn initial(ty: Type) -> Self {
+        Self {
+            ty,
+            option: false,
+            vec: false,
+            result: false,
```

Comment:
Would it mess you up to use the more conventional `is_` or `has_` prefix on these?

Reply:
Nope, gc 👍